### PR TITLE
Stop command to have flagship perform a full stop

### PR DESCRIPTION
**Pull request by petervdmeer**
**Feature:** This PR implements the feature request discussed in issue #4135 (and in other tickets referenced from #4135)

## Feature Details
This PR moves reading of the Shift-input key from AI.cpp to Command.cpp (where all other keyboard inputs for flight are also read). This PR also adds a full-stop command for the flagship using the shift key + "Back" command input.

## UI Screenshots
N/A

## Usage Examples
Fly around, then press shift+reverse to stop your ship.

### CODE REVIEW

**Comment by NomadicVolcano** [comment]
My pull request #4756 will conflict with this (I need to add awareness of full stop).  Can someone please confirm that #4755 will go in?  I will update my pull request with these changes.

**Comment by NomadicVolcano** [comment]
I tried this change, but I found it very error-prone.  More often than not, I ended up telling my fleet to hold position.  Also, shift and H are far apart on the keyboard.

I suggest shift-BACK instead.  I'm going to do a pull request from my branch to petervdmeer's branch with this change.

**Comment by Amazinite** [comment]
>Also, shift and H are far apart on the keyboard. I suggest shift-BACK instead.

All controls are configurable. H is just mentioned because that's the default hold command key. Said more correctly, it's `shift+<hold command>` that causes the flagship to hold, not just shift+H.

**Comment by NomadicVolcano** [comment]
Whatever you map it to, you'll still end up holding your fleet if you release shift before you release HOLD.  I tried shift-BACK, and it works much better.

An alternative is to add a new key for "STOP."  I prefer shift-BACK since it is quick enough and intuitive enough that it is almost reflex.

Get the "down_stop_command" branch from https://github.com/NomadicVolcano/endless-sky to try out my version.  I don't want to make a competing pull request since that will just make the discussion hard to follow.

**Comment by petervdmeer** [comment]
> My pull request #4756 will conflict with this (I need to add awareness of full stop). Can someone please confirm that #4745 will go in? I will update my pull request with these changes.

Nice of you to consider merge-conflicts, but don't worry too much about them. If your change gets merged, then I can also handle the resulting merge-conflict.

> I suggest shift-BACK instead.

Yes, shift-BACK would actually be nicer than shift-H, since BACK is already a ship movement key while the H-key is a fleet command key (and it is indeed also quite intuitive).

When implementing this change then I also considered the BACK-key (without shift). The normal behavior of the BACK-key for a ship without reverse thrusters is that it turns the ship against the movement direction. Adding a second piece of functionality to execute STOP after the turn is complete (as long as BACK is pressed) would allow to implement the full stop without the auto-pilot.

> I'm going to do a pull request from my branch to petervdmeer's branch with this change.

That is fine, but if you create a better implementation, then I would also be okay with just closing this pull-request.

**Comment by NomadicVolcano** [comment]
> That is fine, but if you create a better implementation, then I would also be okay with just closing this pull-request.

I did a pull request to your branch in your fork of the repository.  If you merge that pull request, it will automatically update your pull request here.  

If you want me to make a new pull request directly to endless-sky/endless-sky/master instead, then please respond to this comment explicitly requesting that I do that.

**Comment by petervdmeer** [comment]
> My pull request #4756 will conflict with this (I need to add awareness of full stop). Can someone please confirm that #4745 will go in? I will update my pull request with these changes.

Merged master (and performed conflict resolution) to fix merge-conflict.

I also tested the change to go for Shift + Back. It indeed works quite intuitively.

File: source/AI.cpp [hunk 1]
```diff
@@ -278,14 +276,14 @@ void AI::IssueMoveTarget(const PlayerInfo &player, const Point &target, const Sy
 // Commands issued via the keyboard (mostly, to the flagship).
 void AI::UpdateKeys(PlayerInfo &player, Command &activeCommands, bool isActive)
 {
-	shift = (SDL_GetModState() & KMOD_SHIFT);
```
Comment (tehhowch) [inline source/AI.cpp]: You could simplify the PR diff by defining shift instead of removing it.

File: source/Engine.cpp [hunk 2]
```diff
@@ -1625,8 +1625,22 @@ void Engine::HandleKeyboardInputs()
...
+	// Translate shift+BACK to a command to a STOP command to stop all movement of the flagship.
+	// Translation is done here to allow the autoPilot (which will execute the STOP-command) to
+	// act on a single STOP command instead of the BACK+SHIFT modifier).
+	if(keyHeld.Has(Command::BACK) && keyHeld.Has(Command::SHIFT))
+		activeCommands |= Command::STOP;
+	else if(keyHeld.Has(Command::BACK))
+		activeCommands |= Command::BACK;
+	
+	// Transfer all commands that need to be active as long as the corresponding key is pressed.
+	activeCommands |= (keyHeld.And(Command::PRIMARY | Command::SECONDARY | Command::SCAN |
```
Comment (tehhowch) [inline source/Engine.cpp]: I think this line is partially duplicating what is done in lines 1601-1603

File: source/Engine.cpp [hunk 3]
```diff
@@ -1625,8 +1625,22 @@ void Engine::HandleKeyboardInputs()
 	else
 		activeCommands.Clear(Command::WAIT);
 	
-	// Transfer all newly pressed, unhandled keys to active commands.
-	activeCommands |= keyDown;
+	// Translate shift+BACK to a command to a STOP command to stop all movement of the flagship.
```
Comment (tehhowch) [inline source/Engine.cpp]: You use both "shift+BACK" and "BACK+SHIFT" in this comment. (Also, "autoPilot" is a variable name, while "autopilot" is a concept whose implementation in ES uses a variable named autoPilot.)

File: source/Engine.cpp (hunk 2 above)
Comment (petervdmeer) [reply source/Engine.cpp]: Agreed. I merged the two statements and did a short re-test. The movement keys and stop command still appear to be working fine after merging those.

File: source/Engine.cpp [hunk 4]
```diff
@@ -1625,8 +1630,17 @@ void Engine::HandleKeyboardInputs()
...
+	// Translation is done here to allow the autoPilot (which will execute the STOP-command) to
+	// act on a single STOP command instead of the shift+BACK modifier).
+	if(keyHeld.Has(Command::BACK) && keyHeld.Has(Command::SHIFT))
```
Comment (tehhowch) [inline source/Engine.cpp]: Alternately, we could keep the current `| manueveringCommands);` bit, and then use `activeCommands.Clear(Command::BACK)` if BACK needs to be replaced with STOP. That might result in clearer logic.

File: source/AI.cpp [hunk 5]
```diff
@@ -3101,7 +3099,7 @@ void AI::MovePlayer(Ship &ship, const PlayerInfo &player, Command &activeCommand
 				// if the player is repeatedly targeting nearest to, say, target
 				// a bunch of fighters, they won't start firing on friendly
 				// ships as soon as the last one is gone.
-				if((!state && !shift) || other->IsYours())
+				if((!state && !activeCommands.Has(Command::SHIFT)) || other->IsYours())
```
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/AI.cpp]: I just changed this to have a local variable called shift. I also don't know a better name.

File: source/Engine.cpp [hunk 6]
```diff
@@ -1600,7 +1600,12 @@ void Engine::HandleKeyboardInputs()
 	// Certain commands are always sent when the corresponding key is depressed.
 	static const Command manueveringCommands = Command::AFTERBURNER | Command::BACK |
 		Command::FORWARD | Command::LEFT | Command::RIGHT;
-	activeCommands |= keyHeld.And(Command::PRIMARY | Command::SECONDARY | Command::SCAN | manueveringCommands);
+	
+	// Transfer all commands that need to be active as long as the corresponding key is pressed.
+	// BACK has a different meaning depending on if shift was pressed (and is handled separately).
+	activeCommands |= (keyHeld.And(Command::PRIMARY | Command::SECONDARY | Command::SCAN |
```
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/Engine.cpp]: Nice suggestion, but it actually didn't look good (the AndNot within an And block). I used your later suggestion.

File: source/Engine.cpp (hunk 4 above)
Comment (petervdmeer) [reply source/Engine.cpp]: Yes, this is the alternative I took.

File: source/Engine.cpp [hunk 7]
```diff
@@ -1621,12 +1624,19 @@ void Engine::HandleKeyboardInputs()
 	// If holding JUMP or toggling LAND, also send WAIT. This prevents the jump from
 	// starting (e.g. while escorts are aligning), or switches the landing target.
 	if(keyHeld.Has(Command::JUMP) || (keyHeld.Has(Command::LAND) && landKeyInterval < landCooldown))
-		activeCommands.Set(Command::WAIT);
-	else
-		activeCommands.Clear(Command::WAIT);
+		activeCommands |= Command::WAIT;
 	
-	// Transfer all newly pressed, unhandled keys to active commands.
+	// Transfer all newly pressed unhandled keys to active commands, except for BACK, since
```
Comment (tehhowch) [inline source/Engine.cpp]: What's wrong with the original comment here?

(Also, `autoPilot` in the comment below should really be autopilot and not the name of a variable.)

Comment (NomadicVolcano) [reply source/Engine.cpp]: I think that was my doing, from the pull request that petervdmeer merged into his branch.  I have no strong opinions about the exact wording and capitalization of a comment.

Comment (petervdmeer) [reply source/Engine.cpp]: Doesn't matter a lot who did what. I fixed both comments in the latest commit.

**Comment by tehhowch** [comment]
There should probably be a message that indicates why your ship is suddenly moving by itself. Note that the "Disengaging autopilot." message will appear if the player presses Shift + BACK and then some other canceler key.

**Comment by petervdmeer** [comment]
> There should probably be a message that indicates why your ship is suddenly moving by itself. 

Good idea, I just added it.

**Comment by tehhowch** [comment]
Note that "hold position" and "stop" are different. (Adding a "hold position" for the flagship is

**Comment by petervdmeer** [comment]
> Note that "hold position" and "stop" are different. (Adding a "hold position" command for the player's flagship is not going to merge, but a command to execute a one-time stop is OK.)

This PR implements the one-time stopping action for the players flagship. I just thought that the word "stop" was confusing, since it can refer to other things than movement. (Stop charging shields, stop charging power.)

The message is now updated to use standstill, since standstill seems quite focused on the  removal/absence of movement. I'm not a native American English speaker, so let me know if you prefer a different message.

**Comment by beccabunny** [comment]
An option for the dialog may be "Reaching full stop."

**Comment by petervdmeer** [comment]
Other options which I can think of are:
- Stopping.
- Engaging autopilot to come to full stop.

**Comment by tehhowch** [comment]
I looked at the other messages that autopilot commands generate:

 - `"Landing on ____"` (first press of L)
 - `"You can land on more than one .... Landing on ____"` (first press of L with multiple targets present)
 - `"Switching landing targets. Now landing on ____"` (second press of L with multiple targets)
 - `"Autopilot: landing on ____"` (entered a system with a planet set by map routing)
 - `"Engaging autopilot to jump to the ____ system"` (any press of J)
 - `"Landing on a local wormhole to navigate to the ____ system"` (press of J when wormhole would be used)

Given this, we don't really have a standard format for autopilot messages, and something concise is preferable, I think `"Coming to a stop."` is probably best
